packages/openactive-integration-tests/test/helpers/flow-helper.js

```diff
@@ -79,9 +87,40 @@ class FlowHelper {
     return this.state.putOrder();
   }, { cachePromiseRejection: true });
 
+  P = pMemoize(async () => {
+    this._assertStageShouldNotBeSkipped('P');
+    this.getOrderAfterPPromise = this.state.getOrderAfterP();
+
+    if (!this._stagesToSkip.has('C2')) {
+      await this.C2();
+      if (!this.state.C2ResponseReceived) throw Error('Pre-requisite step failed: C2 failed');
+    }
+
+    return this.state.putOrderProposal();
+  }, { cachePromiseRejection: true });
+
+  getFeedUpdateAfterP = pMemoize(async () => {
+    this._assertStageShouldNotBeSkipped('getFeedUpdateAfterP');
+    if (!this._stagesToSkip.has('P')) {
+      await this.P();
+      if (!this.state.PResponseSucceeded) throw Error('Pre-requisite step failed: P failed');
+    }
+
+    return await this.getOrderAfterPPromise;
+  }, { cachePromiseRejection: true });
+
+  BAfterP = pMemoize(async () => {
```

Comment:
@nickevansuk I think we could do with rethinking the flow-helper logic so that you define specific flows. e.g. we can have the default flow be `C1 -> C2 -> B`, as well as a cancellation flow which is `(default flow) -> U -> getFeedUpdate`

This way, we don't have to have things like `BAfterP` and `getFeedUpdateAfterU` and the `_stagesToSkip`. It would require a bit of work though e.g.  it might be helpful to have the "flow item" store its own state rather than request-state just having a `BResponse`.. 🤔

Reply:
This would make it easier to slot actions in also. I wrote the generic `awaitStage` function out of the dread of thinking of having to write a custom bit of flow logic for each action's use case in its given flow but it could cause issues in other kinds of flows

Reply:
Is this primarily about figuring out which flow stage waits on what? E.g. that in some cases B depends on C2,  getFeedUpdate depends on U, etc?

I wonder given the number of potential variants mean we should have defaults and "custom", where custom is declares at the top of the test, perhaps as a simple array of stages? We could then define some constants which are just pre-defined arrays e.g. `['C1', 'C2', 'B']`, `['C1', 'C2', 'B', `U`, `getFeedUpdate`]` ?

Perhaps B could then look within the array to figure out what's before it that it should be waiting on?

We also need some way of it knowing that any "get from feed" type of stage need to start before the previous stage (to collect whatever result from it), perhaps we just do that by convention? So `get*` will always start before the previous stage?)

Reply:
Also some tests are `['GetMatch', 'C1', 'C2', 'B', 'GetMatch']`